src/generic.jl

```diff
@@ -185,8 +281,22 @@ function pairwise!(r::AbstractMatrix, metric::PreMetric, a::AbstractMatrix;
     end
 end
 
+"""
+    pairwise!(r::AbstractMatrix, metric::PreMetric, a, b)
+    pairwise!(r::AbstractMatrix, metric::PreMetric, a)
+
+Compute distances between each element of collection `a` and each element of
+collection `b` according to distance `metric`, and store the result in `r`.
+If a single iterable `a` is provided, compute distances between its elements.
+
+`r` must be a matrix with size `length(a) × length(b)`.
+"""
+pairwise!(r::AbstractMatrix, metric::PreMetric, a, b) = _pairwise!(r, metric, a, b)
+pairwise!(r::AbstractMatrix, metric::PreMetric, a) = _pairwise!(r, metric, a)
+
 """
     pairwise(metric::PreMetric, a::AbstractMatrix, b::AbstractMatrix=a; dims)
+    pairwise(metric::PreMetric, a::AbstractMatrix; dims)
 
 Compute distances between each pair of rows (if `dims=1`) or columns (if `dims=2`)
 in `a` and `b` according to distance `metric`. If a single matrix `a` is provided,
```

Comment:
Something to note is that these methods are inconsistent with the ones treating `a` and `b` as iterators of columns: a matrix of vectors will be treated differently from a vector of the same vectors. That's probably OK in practice, but that's one of the reasons why I'd like to move to requiring explicitly writing `pairwise(d, eachol(a), eachcol(b))` in the longer term. That way we won't need `dims` anymore.

Reply:
I'm not exactly sure I understand the inconsistency, actually. Could you please sketch an application case?

Reply:
For example, `pairwise(d, [a, b, c, d])` vs. `pairwise(d, reshape([a, b, c, d], 2, 2))` with `a`, `b`, `c` and `d` vectors of numbers.

Reply:
Ok, then I understood you correctly. Out of the two calls you mentioned, only the first one works. The second one fails because it treats `a`, `b`, `c` and `d` like numbers, but then calls like `abs(a)` (or whatever necessary) fail.

Reply:
Yes. But if we were fully consistent, the second call would be equivalent to the first one, since matrices are just one kind of iterator.